### PR TITLE
[Bandwidth-Limit] Add response trailers for bandwidth delay

### PR DESCRIPTION
**Pull request by zhiyong-gayang**
Commit Message: 
- Add response trailers for the bandwidth limit filter delays.
- Add new metric request_enforced and response_enforced.
- Change the following metrics type from Gauge to Counter (request_incoming_size, response_incoming_size, request_allowed_size, response_allowed_size) to better calculate the network bytes per second.

Additional Description: This is useful when downstream want to understand how much delays that is caused by bandwidth limit filter.

Risk Level: Low
Testing: UT added.
Docs Changes: Updated.
Release Notes: Added.
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: source/extensions/filters/http/bandwidth_limit/bandwidth_limit.cc [hunk 1]
```diff
@@ -172,9 +209,18 @@ void BandwidthLimiter::updateStatsOnDecodeFinish() {
 
 void BandwidthLimiter::updateStatsOnEncodeFinish() {
   if (response_latency_) {
+    const auto& config = getConfig();
+
+    auto response_duration = response_latency_.get()->elapsed().count();
```
Comment (yanavlasov) [inline source/extensions/filters/http/bandwidth_limit/bandwidth_limit.cc:214]: We are planning to make stats write only, except for stat sinks. This code will end up broken in this case. Can a timestamp be used in this case?

File: api/envoy/extensions/filters/http/bandwidth_limit/v3alpha/bandwidth_limit.proto [hunk 2]
```diff
@@ -67,4 +67,12 @@ message BandwidthLimit {
...
   // to enabled.
   config.core.v3.RuntimeFeatureFlag runtime_enabled = 5;
+
+  // Optional the prefix for the response trailers of bandwidth decode/encode delays for stream transfer.
+  // If not set, use the default value "bandwidth-request-delay-ms" and "bandwidth-response-delay-ms".
+  // If set, the trailer name will be set as:
+  // request: response_trailer_prefix + "-bandwidth-request-delay-ms"
+  // response: response_trailer_prefix + "-bandwidth-response-delay-ms"
+  // If EnableMode is DISABLED or REQUEST or delay time = 0, the trailer will not be set.
+  string response_trailer_prefix = 6;
```
Comment (yanavlasov) [inline api/envoy/extensions/filters/http/bandwidth_limit/v3alpha/bandwidth_limit.proto:77]: Do you want to make these trailers optional, such that an operator can enable or disable it?

File: source/extensions/filters/http/bandwidth_limit/bandwidth_limit.h [hunk 3]
```diff
@@ -142,6 +148,8 @@ class BandwidthLimiter : public Http::StreamFilter, Logger::Loggable<Logger::Id:
   std::unique_ptr<Envoy::Extensions::HttpFilters::Common::StreamRateLimiter> response_limiter_;
   Stats::TimespanPtr request_latency_;
   Stats::TimespanPtr response_latency_;
+  uint64_t request_duration_ = 0;
```
Comment (yanavlasov) [inline source/extensions/filters/http/bandwidth_limit/bandwidth_limit.h:151]: Would you want to use `std::chrono::milliseconds` to make units explicit?

**Review by yanavlasov — CHANGES_REQUESTED** [review]
/wait

File: api/envoy/extensions/filters/http/bandwidth_limit/v3alpha/bandwidth_limit.proto (hunk 2 above)
Comment (garyyang2002) [reply api/envoy/extensions/filters/http/bandwidth_limit/v3alpha/bandwidth_limit.proto]: make sense, added property to enable it.

File: source/extensions/filters/http/bandwidth_limit/bandwidth_limit.cc [hunk 4]
```diff
@@ -16,14 +16,29 @@ namespace Extensions {
...
     : runtime_(runtime), time_source_(time_source), enable_mode_(config.enable_mode()),
       limit_kbps_(PROTOBUF_GET_WRAPPED_OR_DEFAULT(config, limit_kbps, 0)),
       fill_interval_(std::chrono::milliseconds(PROTOBUF_GET_MS_OR_DEFAULT(
           config, fill_interval, StreamRateLimiter::DefaultFillInterval.count()))),
       enabled_(config.runtime_enabled(), runtime),
-      stats_(generateStats(config.stat_prefix(), scope)) {
+      stats_(generateStats(config.stat_prefix(), scope)),
+      request_delay_trailer_(config.response_trailer_prefix().empty()
+                                 ? DefaultRequestDelayTrailer
+                                 : Http::LowerCaseString(config.response_trailer_prefix() + "-" +
```
[the remark this answers is not in the document]
Comment (garyyang2002) [reply source/extensions/filters/http/bandwidth_limit/bandwidth_limit.cc:36]: strcat seems can't concatenate two const strings, maybe another way I don't know.

File: source/extensions/filters/http/bandwidth_limit/bandwidth_limit.cc (hunk 1 above)
Comment (garyyang2002) [reply source/extensions/filters/http/bandwidth_limit/bandwidth_limit.cc]: To get from the stats will give me a consistent delay value for response trail and the stats. 
Make stats write only seems a challenge work, can we read the stats here before you have a concrete plan how to remove the stats read reference.

File: source/extensions/filters/http/bandwidth_limit/bandwidth_limit.h (hunk 3 above)
Comment (garyyang2002) [reply source/extensions/filters/http/bandwidth_limit/bandwidth_limit.h:153]: the source of request_duration_ is uint64, and this variable will be write to response trailer as string, use milliseconds will need additional convert from unit64 to milliseconds, then to string.

File: api/envoy/extensions/filters/http/bandwidth_limit/v3alpha/bandwidth_limit.proto [hunk 5]
```diff
@@ -67,4 +67,17 @@ message BandwidthLimit {
   // Runtime flag that controls whether the filter is enabled or not. If not specified, defaults
   // to enabled.
   config.core.v3.RuntimeFeatureFlag runtime_enabled = 5;
+
+  // Optional flag to enabled response trailers.
+  // Default is false.
```
Comment (htuch) [inline api/envoy/extensions/filters/http/bandwidth_limit/v3alpha/bandwidth_limit.proto:72]: Default and optional are redundant for scalars, so just say `Enable response trailers` and link to documentation or explain what they are.

Comment (htuch) [reply api/envoy/extensions/filters/http/bandwidth_limit/v3alpha/bandwidth_limit.proto:72]: Also, single or plural response trailers?

File: api/envoy/extensions/filters/http/bandwidth_limit/v3alpha/bandwidth_limit.proto [hunk 6]
```diff
@@ -67,4 +67,17 @@ message BandwidthLimit {
...
+  bool enable_response_trailer = 6;
+
+  // Optional the prefix for the response trailers of bandwidth decode/encode delays for stream transfer.
```
Comment (htuch) [inline api/envoy/extensions/filters/http/bandwidth_limit/v3alpha/bandwidth_limit.proto:75]: I don't really understand what this feature is or how it works from the API docs. Can you either expand them or add some explanation in the main Envoy docs and link there?

**Review by htuch — CHANGES_REQUESTED** [review]
/wait

File: api/envoy/extensions/filters/http/bandwidth_limit/v3alpha/bandwidth_limit.proto (hunk 6 above)
Comment (garyyang2002) [reply api/envoy/extensions/filters/http/bandwidth_limit/v3alpha/bandwidth_limit.proto]: Move to the description to `enable_response_trailers` where seems better place for the explanation.

File: api/envoy/extensions/filters/http/bandwidth_limit/v3alpha/bandwidth_limit.proto (hunk 5 above)
Comment (garyyang2002) [reply api/envoy/extensions/filters/http/bandwidth_limit/v3alpha/bandwidth_limit.proto]: updated, try to make it more clear.

File: docs/root/version_history/current.rst [hunk 7]
```diff
@@ -32,6 +32,8 @@ Minor Behavior Changes
 ----------------------
 *Changes that may cause incompatibilities for some users, but should not for most*
 
+* bandwidth_limit: added response trailers when request or response delay are enforced.
```
Comment (htuch) [inline docs/root/version_history/current.rst:35]: Please add RST links to the API fields here and below.

File: api/envoy/extensions/filters/http/bandwidth_limit/v3alpha/bandwidth_limit.proto [hunk 8]
```diff
@@ -67,4 +67,19 @@ message BandwidthLimit {
   // Runtime flag that controls whether the filter is enabled or not. If not specified, defaults
   // to enabled.
   config.core.v3.RuntimeFeatureFlag runtime_enabled = 5;
+
+  // Optional Flag to enable response trailers.
```
Comment (htuch) [inline api/envoy/extensions/filters/http/bandwidth_limit/v3alpha/bandwidth_limit.proto:71]:
```suggestion
  // Enable response trailers.
```

File: api/envoy/extensions/filters/http/bandwidth_limit/v3alpha/bandwidth_limit.proto [hunk 9]
```diff
@@ -67,4 +67,19 @@ message BandwidthLimit {
...
+  //
+  // .. note::
+  //   If set true, the response trailer "bandwidth-request-delay-ms" and "bandwidth-response-delay-ms"
```
Comment (htuch) [inline api/envoy/extensions/filters/http/bandwidth_limit/v3alpha/bandwidth_limit.proto:74]:
```suggestion
  //   If set true, the response trailers *bandwidth-request-delay-ms* and *bandwidth-response-delay-ms*
```

File: api/envoy/extensions/filters/http/bandwidth_limit/v3alpha/bandwidth_limit.proto [hunk 10]
```diff
@@ -67,4 +67,19 @@ message BandwidthLimit {
...
+  //   will be added.
```
Comment (htuch) [inline api/envoy/extensions/filters/http/bandwidth_limit/v3alpha/bandwidth_limit.proto:75]:
```suggestion
  //   will be added, prefixed by *response_trailer_prefix*.
```

File: api/envoy/extensions/filters/http/bandwidth_limit/v3alpha/bandwidth_limit.proto [hunk 11]
```diff
@@ -67,4 +67,19 @@ message BandwidthLimit {
...
+  //   bandwidth-request-delay-ms: delay time in milliseconds it took for the request stream transfer.
```
Comment (htuch) [inline api/envoy/extensions/filters/http/bandwidth_limit/v3alpha/bandwidth_limit.proto:76]: Make these RST lists, I suggest looking at docs rendering preview for the API docs for this filter.

File: api/envoy/extensions/filters/http/bandwidth_limit/v3alpha/bandwidth_limit.proto [hunk 12]
```diff
@@ -67,4 +67,19 @@ message BandwidthLimit {
...
+  //   bandwidth-response-delay-ms: delay time in milliseconds it took for the response stream transfer.
+  //   If enable_mode is DISABLED or REQUEST, the trailers will not be set.
```
Comment (htuch) [inline api/envoy/extensions/filters/http/bandwidth_limit/v3alpha/bandwidth_limit.proto:78]: RST link to `enbable_mode`.

**Review by htuch — CHANGES_REQUESTED** [review]
Much clearer comments, a bunch of RST nits.
/wait

**Review by htuch — APPROVED** [review]
LGTM, thanks!

**Pull request merged**